codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/EndpointHostPrefixMiddleware.java

```diff
@@ -0,0 +1,204 @@
+
+                String middlewareName = getMiddlewareName(operation);
+                String middlewareHelperName = getMiddlewareHelperName(operation);
+                writer.openBlock("func $L(stack *middleware.Stack) error {", "}",
+                        middlewareHelperName,
+                        () -> {
+                            writer.write(
+                                    "return stack.Serialize.Insert(&$L{}, `OperationSerializer`, middleware.Before)",
+                                    middlewareName);
+                        });
+
+            });
+        });
+    }
+
+    private static void writeMiddleware(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            OperationShape operation,
+            SmithyPattern pattern
+    ) {
+        GoStackStepMiddlewareGenerator middlewareGenerator =
+                GoStackStepMiddlewareGenerator.createInitializeStepMiddleware(
+                        getMiddlewareName(operation),
+                        MIDDLEWARE_ID
+                );
+
+        middlewareGenerator.writeMiddleware(writer, (generator, w) -> {
+            writer.addUseImports(SmithyGoDependency.SMITHY_HTTP_TRANSPORT);
+            writer.addUseImports(SmithyGoDependency.FMT);
+
+            w.openBlock("if smithyhttp.GetHostnameImmutable(ctx) {", "}", () -> {
+                w.write("return next.$L(ctx, in)", generator.getHandleMethodName());
+            }).write("");
+
+            w.write("req, ok := in.Request.(*smithyhttp.Request)");
+            w.openBlock("if !ok {", "}", () -> {
+                writer.write("return out, metadata, fmt.Errorf(\"unknown transport type %T\", in.Request)");
+            }).write("");
+
+            if (pattern.getLabels().isEmpty()) {
+                w.write("request.HostPrefix = $S", pattern.toString());
+
+            } else {
+                // If the pattern has labels, we need to build up the host prefix using a string builder.
+                writer.addUseImports(SmithyGoDependency.STRINGS);
+
+                w.write("var prefix strings.Builder");
+                StructureShape input = ProtocolUtils.expectInput(model, operation);
+                for (SmithyPattern.Segment segment : pattern.getSegments()) {
+                    if (!segment.isLabel()) {
+                        w.write("prefix.WriteString($S)", segment.toString());
+                    } else {
+                        MemberShape member = input.getMember(segment.getContent()).get();
+                        String memberName = symbolProvider.toMemberName(member);
+                        String memberReference = "input." + memberName;
+
+                        // Theoretically this should never be nil or empty by this point unless validation has
+                        // been disabled.
+                        w.write("if $L == nil || len(*$L) == 0 {", memberReference, memberReference).indent();
```

Comment:
QQ: Are we certain host prefix would not be numeric?

Reply:
Correct, host prefix labels must only be string values. But I've also removed this `len` check as its duplicates the validate method in the if else case.